packages/react-storage/src/components/StorageBrowser/views/hooks/useSearch.ts

```diff
@@ -0,0 +1,77 @@
+import React from 'react';
+import debounce from 'lodash/debounce';
+
+const DEFAULT_DELAY = 300;
+export interface InitialSearchValues<T> {
+  searchKey: keyof T;
+  searchTerm?: string;
+  debounceDelay?: number;
+}
+
+interface UseSearchProps<T> {
+  initialValues: InitialSearchValues<T>;
+  items: T[];
+  onSearch: (term: string, includeSubfolders?: boolean) => void;
+}
+
+interface UseSearch<T> {
+  filteredItems: T[];
+  searchTerm: string;
+  handleSearch: (term: string, includeSubfolders?: boolean) => void;
+}
+
+/**
+ * Custom hook to search a list of items based on a search key and search term.
+ * @template T - The type of items in the search list.
+ * @param {UseSearchProps<T>} props
+ * @returns {UseSearch<T>} - The search results, current term, and handler.
+ */
+export function useSearch<T>({
+  items,
+  initialValues: initialValuesProp,
+  onSearch,
+}: UseSearchProps<T>): UseSearch<T> {
+  const [initialValues] = React.useState(initialValuesProp);
```

Comment:
Is this needed? Afaict, it seems like we can just have `initialValues` be `values` in the hook param and just always use below.

Reply:
The intention with `initialValues` is to keep it stable for the lifecycle of a component as they're just initial values as the names suggest. e.g after the initial render, this hook is the source of truth for what the current `searchTerm` and it manages that internally (with debouncing for e.g) . We don't want to let the component change that after the first render.

Reply:
Prefer `useRef` for these scenarios